when tagpool is exhausted, grow up to RFC 6 limits

garlick opened the pull request:
Rather than failing at 64K matchtags, allow the non-group tagpool to grow up to 1M in size.

Also: change kvs_watch to use the non-group tagpool to hopefully address wreck scalability issues discussed in #802.

This took a while due to some disturbing behavior by libveb where initialization to "allocated" did not work, depending on the size of the veb tree - this is why I've added so many assertions to the code.  I did verify that the old implementation worked as it should.  In fact this condition was caught by an existing unit test that tries to run out the pool and verify that it really got all the tags.  I couldn't find any memory clobbering with valgrind, so I assume this is an internal bug in libveb, and it doesn't seem to occur when the tree is initialized to "not allocated".  That is how it's used in nodeset.c and here now as well.

trws commented:
Interesting about veb... hope that resolved ok.  The update here looks good, but one thing we talked about the other day that might be useful is to have a diagnostic when the tag pool is forced to grow over a given threshold, actually over 4096 seems reasonable, so we can tell when something fishy is going on.  Do you think it would be reasonable to include something like that here?

garlick commented:
Sorry I submitted this PR prematurely.  I've gone down a rat hole debugging this, as it still seems not to be working.

I'll ponder the diagnostic - it's a bit tricky figuring how to do it in a way that works for all contexts (command, broker, module, etc).

garlick commented:
OK this is finally getting close I think.

I fixed a bug in libveb where `vebnew()` was improperly setting up the veb tree when its "full" parameter was set.  I should use the term "fixed" loosely - I just made it handle that case by creating the tree "empty", then walking the tree setting each bit.  Will get a bug open upstream, although the code hasn't changed since 2010.

In the process of running this down, I pulled in the upstream veb unit tests and added some more including tests for the failing case fixed above.

I also fixed a problem with kvs.unwatch handling, introduced in pr #788.

Having fixed those things, changing the kvs to the non-group matchtag pool, and allowing the non-group pool to grow to the full size allowed by RFC 6 (1M) works.  I added a test that checks how many kvs watches can be started on one handle, and ran it up to 402589 before my cloud9 VM ran out of memory.  I added this to travis-ci at 8192 watchers, just to prove it goes past the previous limit of 4096.

garlick commented:
On the diagnostic, writing to stderr is likely to be a problem in some contexts (for example if stdio has been closed and the fd's reused).  Would calling flux_log at LOG_CRIT each time we double the size of the tagpool be useful?

trws commented:
That sounds reasonable to me.

garlick commented:
OK, I made the logging level LOG_INFO though.  LOG_CRIT is probably overkill.  At least if things are going slowly a little clue may be waiting for you in the logs, e.g.

```
2016-09-13T22:33:05.345695Z -.info[0]: tagpool-normal expanded from 1024 to 2048 entries
2016-09-13T22:33:05.712806Z -.info[0]: tagpool-normal expanded from 2048 to 4096 entries
2016-09-13T22:33:06.829736Z -.info[0]: tagpool-normal expanded from 4096 to 8192 entries
```

grondo commented:
> OK, I made the logging level LOG_INFO though. LOG_CRIT is probably overkill. At least if things are going slowly a little clue may be waiting for you in the logs, e.g.

I think that is perfect, I was also going to suggest that LOG_INFO is adequate.

Is there any way to tag the log message coming from an API user with process name or UUID so when/if we do see one of the messages we can guess at the source?

One idea for scalability testsuite would be to check or filter for these messages after runs.

garlick commented:
I was just wondering about the "-" in the log message.  That is the default "appname" if  `flux_log_set_appname()` has not been called.   I had put together a quick fix to make the pid the default; then I thought I'd better look a little more carefully to see why the pid isn't already in the logs since it's part of the message.

Actually, making the default appname the process name as a default is an excellent idea.

garlick commented:
Easier than I thought:
```
2016-09-13T22:49:21.625546Z lt-watch.info[0]: tagpool-normal expanded from 1024 to 2048 entries
2016-09-13T22:49:21.900167Z lt-watch.info[0]: tagpool-normal expanded from 2048 to 4096 entries
2016-09-13T22:49:22.465946Z lt-watch.info[0]: tagpool-normal expanded from 4096 to 8192 entries
2016-09-13T22:49:23.627800Z lt-watch.info[0]: tagpool-normal expanded from 8192 to 16384 entries
```

garlick commented:
This is ready to be considered for a merge.

grondo commented:
Really nice work here @garlick! I'll poke at it but it looks great to me.

grondo commented:
This is perfect. Merging.

garlick commented:
Thanks!

On Sep 14, 2016 8:19 AM, "Mark Grondona" <notifications@github.com> wrote:

> Merged #806 <https://github.com/flux-framework/flux-core/pull/806>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/806#event-789306160>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKX2w4uEbaIjBs6sSEiNu6fYDoQW8miks5qqBCagaJpZM4J7Sub>
> .
>
